halo/hooks/object.cpp

```diff
@@ -28,9 +32,240 @@ Signature(true, sig_biped_jump,
+ObjectBehaviorDefinition new_unit_beh;
+
+ObjectBehaviorDefinition new_bipd_beh;
+ObjectBehaviorDefinition new_vehi_beh;
+ObjectBehaviorDefinition new_weap_beh;
+
+ObjectBehaviorDefinition new_eqip_beh;
+ObjectBehaviorDefinition new_garb_beh;
+ObjectBehaviorDefinition new_proj_beh;
+
+ObjectBehaviorDefinition new_scen_beh;
+ObjectBehaviorDefinition new_mach_beh;
+ObjectBehaviorDefinition new_ctrl_beh;
+ObjectBehaviorDefinition new_lifi_beh;
+
+ObjectBehaviorDefinition new_plac_beh;
+ObjectBehaviorDefinition new_ssce_beh;
+
+extern "C" {
+
+    // Negative type indices
+
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(object);
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(device);
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(item);
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(unit);
+
+    // Positive type indices
+
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(biped);
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(vehicle);
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(weapon);
+
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(equipment);
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(garbage);
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(projectile);
+
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(scenery);
+
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(machine);
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(control);
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(light_fixture);
+
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(placeholder);
+    OBJECT_BEHAVIOR_FUNCTION_PTRS(sound_scenery);
+
+}
+
+#define COPY_EVENT_DEF(type, def) \
+    type ## _adjust_placement = def.adjust_placement; \
+    type ## _create = def.create; \
+    type ## _place = def.place; \
+    type ## _destroy = def.destroy; \
+    type ## _update = def.update; \
+    type ## _export_function_values = def.export_function_values; \
+    type ## _handle_deleted_object = def.handle_deleted_object; \
+    type ## _handle_region_destroyed = def.handle_region_destroyed; \
+    type ## _handle_parent_destroyed = def.handle_parent_destroyed; \
+    type ## _preprocess_node_orientations = def.preprocess_node_orientations; \
+    type ## _postprocess_node_matricies = def.postprocess_node_matricies; \
+    type ## _reset = def.reset; \
+    type ## _disconnect_from_structure_bsp = def.disconnect_from_structure_bsp; \
+    type ## _notify_impulse_sound = def.notify_impulse_sound; \
+    type ## _render_debug = def.render_debug; \
+    type ## _new_to_network = def.new_to_network; \
+    type ## _update_baseline = def.update_baseline; \
+    type ## _generate_update_header = def.generate_update_header; \
+    type ## _handle_received_update = def.handle_received_update; \
+    type ## _should_do_update_incremental = def.should_do_update_incremental; \
+    type ## _set_last_update_time = def.set_last_update_time; \
+    // End of Macro
+
+static ObjectBehaviorDefinition* backup[12];
 
 void init_object_hooks(){
+    auto game_defs = object_behavior_defs();
+
+    // Copy data to our new defs.
+
+    memcpy(&new_obje_beh, game_defs[static_cast<int>(ObjectType::BIPED)]->parent_definitions[0],  sizeof(ObjectBehaviorDefinition));
```

Comment:
You could shorten all of these lines by storing the value of `sizeof(ObjectBehaviorDefinition)` in a variable